crypto/initthread.c

```diff
@@ -257,7 +257,7 @@ void *ossl_thread_event_ctx_new(OSSL_LIB_CTX *libctx)
     if (tlocal == NULL)
         return NULL;
 
-    if (!CRYPTO_THREAD_init_local(tlocal,  NULL)) {
+    if (!CRYPTO_THREAD_init_local(tlocal, &ossl_thread_event_ctx_free)) {
```

Comment:
Why is this needed?
Doesnt context_deinit_objs() call ossl_thread_event_ctx_free()?

Reply:
Without it, there is a leak: the call comes from the wrong thread and doesn't pick up the thread local.

Reply:
I'm not sure this is quite correct. This can happen *inside* the FIPS provider. This causes the `ossl_thread_event_ctx_free` function to be called on thread destruction. But, if the FIPS provider gets unloaded *before* thread destruction, I think this could lead to a crash? Shouldn't a call to `ossl_thread_event_ctx_free` be added as the last step in the FIPS variant of  `ossl_ctx_thread_stop` (starting on line 288 below) instead?

Reply:
Moving the free there does **not** address the leak.
```diff
diff --git a/crypto/context.c b/crypto/context.c
index 33d52a964b..68710931b3 100644
--- a/crypto/context.c
+++ b/crypto/context.c
@@ -652,3 +652,11 @@ const char *ossl_lib_ctx_get_descriptor(OSSL_LIB_CTX *libctx)
     return "Non-default library context";
 #endif
 }
+
+#ifdef FIPS_MODULE
+void ossl_lib_ctx_fips_thread_cleanup(OSSL_LIB_CTX *ctx)
+{
+    ossl_thread_event_ctx_free(ctx->thread_event_handler);
+    ctx->thread_event_handler = NULL;
+}
+#endif
diff --git a/crypto/initthread.c b/crypto/initthread.c
index 3b281dc2c9..2ec5f20f85 100644
--- a/crypto/initthread.c
+++ b/crypto/initthread.c
@@ -257,7 +257,7 @@ void *ossl_thread_event_ctx_new(OSSL_LIB_CTX *libctx)
     if (tlocal == NULL)
         return NULL;
 
-    if (!CRYPTO_THREAD_init_local(tlocal, &ossl_thread_event_ctx_free)) {
+    if (!CRYPTO_THREAD_init_local(tlocal, NULL)) {
         goto err;
     }
 
@@ -285,6 +285,7 @@ static void ossl_arg_thread_stop(void *arg)
     ossl_ctx_thread_stop((OSSL_LIB_CTX *)arg);
 }
 
+void ossl_lib_ctx_fips_thread_cleanup(OSSL_LIB_CTX *);
 void ossl_ctx_thread_stop(OSSL_LIB_CTX *ctx)
 {
     THREAD_EVENT_HANDLER **hands;
@@ -293,6 +294,7 @@ void ossl_ctx_thread_stop(OSSL_LIB_CTX *ctx)
 
     if (local == NULL)
         return;
+    ossl_lib_ctx_fips_thread_cleanup(ctx);
     hands = init_get_thread_local(local, 0, 0);
     init_thread_stop(ctx, hands);
     OPENSSL_free(hands);
```

Reply:
I'm pretty sure that this is the correct approach.  The handler is being installed on the FIPS provider's _internal_ libctx.  It isn't possible to unload the FIPS provider without cleaning this up.

Reply:
The more I look at this the more I realise this is *definitely wrong*. The function `ossl_thread_event_ctx_free` is intended to free the `ossl_thread_event_ctx` - which is just a thread local *key*. The destructor function that you pass as the second argument to `CRYPTO_THREAD_init_local` gets called when threads are destroyed and is there to free up thread local *values*.

You are confusing the key and the value here.

But this is not the way things are supposed to work. When threads get destroyed, or the user calls `OPENSSL_thread_stop` explicitly, OpenSSL is itself supposed to detect that and call all its internal thread_stop functions automatically. It is supposed to be sufficient to clean up any thread local values in the `ossl_ctx_thread_stop` function. This should be true even in a provider. There are core callbacks that are supposed to register all this stuff.

I just tried to replicate the mem leak but was unable to. I backed out the change in crypto/initthread.c and ran this PR with "enable-asan" but did not see a mem leak reported. Where are you seeing a mem leak and how can I reproduce it?